fileutil/caching_file_reader.go

```diff
@@ -0,0 +1,68 @@
+package fileutil
+
+import (
+	"io/ioutil"
+	"sync"
+	"time"
+)
+
+// CachingFileReader reads a file and keeps an in-memory copy of it, until the
+// copy is considered stale. Next ReadFile() after expiry will re-read the file from disk.
+type CachingFileReader struct {
+	// path is the file path to the cached file.
+	path string
+
+	// ttl is the time-to-live duration when cached file is considered stale
+	ttl time.Duration
+
+	// cache is the buffer holding the in-memory copy of the file.
+	cache cachedFile
+
+	l sync.RWMutex
+
+	// currentTime is a function that returns the current local time.
+	// Normally set to time.Now but it can be overwritten by test cases to manipulate time.
+	currentTime func() time.Time
+}
+
+type cachedFile struct {
+	// buf is the buffer holding the in-memory copy of the file.
+	buf string
+
+	// expiry is the time when the cached copy is considered stale and must be re-read.
+	expiry time.Time
+}
+
+func NewCachingFileReader(path string, ttl time.Duration, currentTime func() time.Time) *CachingFileReader {
+	return &CachingFileReader{
+		path:        path,
+		ttl:         ttl,
+		currentTime: currentTime,
+	}
+}
+
+func (r *CachingFileReader) ReadFile() (string, error) {
+	// Fast path requiring read lock only: file is already in memory and not stale.
+	r.l.RLock()
+	now := r.currentTime()
+	cache := r.cache
+	r.l.RUnlock()
+	if now.Before(cache.expiry) {
+		return cache.buf, nil
+	}
+
+	// Slow path: read the file from disk.
+	r.l.Lock()
+	defer r.l.Unlock()
+
+	buf, err := ioutil.ReadFile(r.path)
+	if err != nil {
+		return "", err
+	}
+	r.cache = cachedFile{
+		buf:    string(buf),
+		expiry: now.Add(r.ttl),
```

Comment:
It's very unlikely, and would require exceedingly high load, but we could theoretically spend an arbitrary amount of time waiting for the write lock at line 55, and so we could end up storing an expiry time that's already in the past, or at least more stale than it should be.

```suggestion
		expiry: r.currentTime().Add(r.ttl),
```

Reply:
Agreed, thanks!